README.md
# Teamleader

A ruby wrapper around Teamleader.eu API v1.

**Please note that this gem is linked to the [deprecated Teamleader API](https://apidocs.teamleader.be/). Teamleader recommends to use the [new Teamleader API](https://developer.teamleader.eu/).**

Some endpoints are still missing. Don't hesitate to open an issue or a PR if you need one of them.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'teamleader'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install teamleader

## Usage

```ruby
require 'teamleader'
teamleader = Teamleader.new('Your API group', 'Your API secret')

```
You can find your API key in your Teamleader account, under Settings > API & Webhooks. API access is available for every Teamleader account.
Note that you need to be admin to access this page.

### Configuration
For Rails application create a configuration file `config/initializers/teamleader.rb`. You can do this anywhere in your application before you make API calls using the gem.

```ruby
Teamleader.configure do |config|
  config.api_group = '45678'
  config.api_secret = '31e56cf3b3c259f56666ba4a6089ee91c3150683611834f7eb2f5a7a4f039a17910f1fa2d65d282e9c344abcf895dad80a89b13af8fe917dfed1e0798c83350c'
end
```

## Available methods
When a method expects a hash as argument, the hash keys have the same name as described in [Teamleader API documentation](http://apidocs.teamleader.be/).

If a required parameter is missing or if the API returns an error, the method will throw an exception.

### General

```ruby
teamleader.get_users
teamleader.get_departments
teamleader.get_tags
```

### Contacts

```ruby
teamleader.add_contact :forename => "John", :surname => "Doe", :email => "john.doe@provider.com"
teamleader.update_contact({:contact_id => 123, :track_changes => 0, :forename => "Johnny"})
teamleader.delete_contact({:contact_id => 123})
teamleader.link_contact_to_company({:contact_id => 123, :company_id => 456, :mode => "link"})
teamleader.get_contacts({:amount => 100, :pageno => 0, :searchby => "John"}) # pagination starts at 0
teamleader.get_contact({:contact_id => 123})
teamleader.get_contacts_by_company({:company_id => 123})
teamleader.get_contact_company_relations({:amount => 10, :pageno => 0})
```

### Companies

```ruby
teamleader.add_company({:name => "Hello World"})
teamleader.get_company({:company_id => 12345})
teamleader.get_companies({:amount => 100, :pageno => 0, :searchby => "hello"}) # pagination starts at 0
```

### Deals

Supported methods are: `add_deal`, `update_deal`, `update_deal_items`, `send_sale_to_client`, `get_deals`, `get_deals_by_contact_or_company`, `get_deals_by_project`, `get_deal`, `get_deal_phase_changes`, `get_all_deal_phase_changes`, `get_deal_phases`, `get_deal_sources`

### Invoices

```ruby
teamleader.add_invoice({
    :contact_or_company => "company",
    :contact_or_company_id => 12345,
    :sys_department_id => 5678,
    :description_1 => "My awesome item",
    :price_1 => 50,
    :amount_1 => 1
})
teamleader.update_invoice_payment_status({:invoice_id => 123, :status => "paid"})
teamleader.get_invoice({:invoice_id => 123})
teamleader.download_invoice_pdf({:invoice_id => 123})
teamleader.get_invoices({:date_from => "01/01/2017", :date_to => "01/05/2017"})
teamleader.get_creditnotes({:date_from => "01/01/2017", :date_to => "01/05/2017"})
teamleader.get_bookkeeping_accounts({:sys_department_id => 134})
```

#### Other supported methods
`book_draft_invoice`, `update_invoice`, `update_invoice_comments`, `delete_invoice`, `get_creditnote`, `get_invoices_by_project`, `get_creditnotes_by_project`, `get_invoices_by_deal`, `add_creditnote`, `download_creditnote_pdf`

### Products

```ruby
teamleader.add_product({
    :name => "My Awesome Product",
    :price => 49.99
})
teamleader.update_product({:product_id => 123, :price => 39.99})
teamleader.get_product({:product_id => 123})
teamleader.delete_product({:product_id => 123})
teamleader.get_products({:amount => 100, :pageno => 0}) # Pagination starts at 0
```

### Projects

```ruby
teamleader.get_project({ :project_id => 123 })
teamleader.get_projects_by_client({
  contact_or_company: 'contact',
  contact_or_company_id: '123',
  deep_search: 1
})

teamleader.add_project({
  :project_name => 123,
  :project_budget => 300,
  :project_responsible_user_id => 123,
  :project_start_date => '12/01/2017',
  :milestone_title => 'milestone title',
  :milestone_budget => 200,
  :milestone_invoiceable => 0,
  :milestone_due_date => '12/03/2017',
  :milestone_responsible_user_id => 123
})
teamleader.update_project({
  :project_id => 123,
  :track_changes => 0,
  :title => "New Project Title"
})
teamleader.get_projects({ :amount => 50, pageno: 0 })
# Pagination starts at 0
```
### Tickets

Supported methods are: `add_ticket`, `update_ticket`, `add_ticket_message`, `get_tickets`, `get_ticket`, `get_ticket_messages`, `get_ticket_message`, `get_ticket_cloud_url`.

### Notes

Supported methods are: `add_note`, `get_notes`

### Files

```ruby
teamleader.get_file_info({ :file_id => '1236412' })

teamleader.download_file({ :file_id => '1236412' })

teamleader.upload_file({
  :object_type => 'company',
  :object_id => '1236412',
  :file_content => 'Base64 encoded version the file',
  :file_name => 'file.rb'
})

teamleader.delete_file({ :file_id => '1236412' })

teamleader.get_files({
  :amount => 100,
  :pageno => 0,
  :object_type => 'contact',
  :object_id => '1234567'
})
# Pagination starts at 0
```

### Callbacks
```ruby
teamleader.add_callback({
  due_date: 1538400423,
  hour: '10h15',
  user_id: '1234',
  for: 'contact',
  for_id: '9876'
})
teamleader.get_calls(amount: 10, pageno: 0)
teamleader.get_call(call_id: '123456')
```

### Subscriptions
Available methods are: `get_subscriptions`, `get_all_subscriptions`, `add_subscription`, `update_subscription`, `delete_subscription`, `get_subscription`, `get_invoices_by_subscription`, `get_related_subscriptions_by_invoice`, `get_subscriptions_by_deal`, `get_subscriptions_by_contact_or_company`

### Custom Fields
Available methods are: `get_custom_fields`, `get_custom_field_info`, `add_custom_field_option`

## License
The Teamleader GEM is released under the MIT License.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

## Credits
This gem is inspired by the [hunterio gem](https://github.com/prospectio/hunterio/) made by Vincenzo Ruggiero

Made with ️❤️️️ by [Javry](https://javry.com) in Belgium.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.17.1] - 2020-04-09
- Fix include

## [0.17.0] - 2020-04-09
- `get_all_subscriptions` method accepts optional parameters (like `selected_customfields` or `searchby`)
- New methods related to subscriptions: `add_subscription`, `update_subscription`, `delete_subscription`, `get_subscription`, `get_invoices_by_subscription`, `get_related_subscriptions_by_invoice`, `get_subscriptions_by_deal`, `get_subscriptions_by_contact_or_company`

## [0.16.0] - 2019-05-10
- New method related to custom fields: `get_custom_fields`, `get_custom_field_info`, `add_custom_field_option`

## [0.15.0] - 2019-05-09
- New method related to credit notes: `download_creditnote_pdf`

## [0.14.0] - 2018-10-01
- Sync with version published on Rubygems

## [0.13.0] - 2018-10-01
### Added
- New methods related to calls: `get_calls`, `get_call`, `add_callback`

## [0.12.0] - 2018-03-28
### Added
- New methods related to subscriptions: `get_subscriptions`, `get_all_subscriptions`

## [0.11.1] - 2018-03-07
### Fixed
- Error when the API responds with an error but no `reason` attribute in the response body.

## [0.11.0] - 2018-02-15
### Added
- New methods related to invoices and credit notes: `book_draft_invoice`, `update_invoice`, `update_invoice_comments`, `delete_invoice`, `get_creditnote`, `get_invoices_by_project`, `get_creditnotes_by_project`, `get_invoices_by_deal`, `add_creditnote`

## [0.10.0] - 2018-01-09
### Added
- All methods related to companies: `add_company`, `update_company`, `delete_company`, `get_companies`, `get_company`, `get_business_types`

## [0.9.0] - 2018-01-05
### Added
- Some methods related to notes: `add_note`, `get_notes`

## [0.8.0] - 2017-11-08
### Added
- Some methods related to invoices: `send_invoice`, `get_invoice_reminder_templates`, `get_invoice_reminder_template_content`, `get_invoice_cloud_url`

## [0.7.0] - 2017-11-08
### Added
- Methods for manipulating Deals

## [0.6.0] - 2017-10-31
### Added
- `get_creditnotes` method

## [0.5.1] - 2017-10-17
### Fixed
- Relax required parameters for `add_ticket` method.

## [0.5.0] - 2017-10-16
### Added
- Add methods for manipulating Tickets

## [0.4.0] - 2017-09-28
### Added
- Add support for getting bookkeeping accounts (cfr `get_bookkeeping_accounts` method)

## [0.3.1] - 2017-05-12
### Fixed
- Parameter check for `link_contact_to_company`

## [0.3.0] - 2017-05-12
### Added
- Methods for Contacts (add, get, update, delete, link to company, search, get contacts from a company, get relationships between contacts and companies)

## [0.2.0] - 2017-05-12
### Added
- Method `link_contact_to_company`
- Methods for Products (add, update, delete, get one, get all)
- Methods for Invoice (add, update payment status, get PDF, get one, get all)
- Update `get_company` to accept an Hash instead of an id as parameter

## [0.1.1] - 2017-05-12
### Changed
- Mark `json` dependency version as `>= 1`

## [0.1.0] - 2017-05-11
### Added
- Initial commit


lib/teamleader/api.rb

require 'json'
require 'net/http'
require 'uri'

module Teamleader
  API_BASE_URL = "https://app.teamleader.eu/api"

  class Api
    include Teamleader::CustomFields
    include Teamleader::Files
    include Teamleader::Projects
    include Teamleader::Calls
    include Teamleader::Subscriptions

    def initialize(group, secret)
      @api_group = group
      @api_secret = secret
      uri = URI.parse(API_BASE_URL)
      @http = Net::HTTP.new(uri.host, uri.port)
      @http.use_ssl = true
    end

    def get_users(params={})
      request "/getUsers.php", params
    end

    def get_departments
      request "/getDepartments.php"
    end

    def get_tags
      request "/getTags.php"
    end

    # Returns the ID of the contact
    def add_contact(params={})
      raise "forename is required" if params[:forename].nil?
      raise "surname is required" if params[:surname].nil?
      raise "email is required" if params[:email].nil?
      request "/addContact.php", params
    end

    def get_contact(params={})
      raise "contact_id is required" if params[:contact_id].nil?
      request "/getContact.php", params
    end

    def update_contact(params={})
      raise "contact_id is required" if params[:contact_id].nil?
      raise "track_changes is required" if params[:track_changes].nil?
      request "/updateContact.php", params
    end

    def delete_contact(params={})
      raise "contact_id is required" if params[:contact_id].nil?
      request "/deleteContact.php", params
    end

    def link_contact_to_company(params={})
      raise "contact_id is required" if params[:contact_id].nil?
      raise "company_id is required" if params[:company_id].nil?
      raise "mode is required" if params[:mode].nil?
      raise "mode must be 'link' or 'unlink'" unless ['link', 'unlink'].include?(params[:mode])
      request "/linkContactToCompany.php", params
    end

    def get_contacts(params={})
      raise "amount is required" if params[:amount].nil?
      raise "pageno is required" if params[:pageno].nil?
      request "/getContacts.php", params
    end

    def get_contacts_by_company(params={})
      raise "company_id is required" if params[:company_id].nil?
      request "/getContactsByCompany.php", params
    end

    def get_contact_company_relations(params={})
      raise "amount is required" if params[:amount].nil?
      raise "pageno is required" if params[:pageno].nil?
      request "/getContactCompanyRelations.php", params
    end

    def add_deal(params={})
      raise "contact_or_company is required" if params[:contact_or_company].nil?
      raise "contact_or_company_id is required" if params[:contact_or_company_id].nil?
      raise "title is required" if params[:title].nil?
      raise "source is required" if params[:source].nil?
      request "/addDeal.php", params
    end

    def update_deal(params={})
      raise "deal_id is required" if params[:deal_id].nil?
      request "/updateDeal.php", params
    end

    def update_deal_items(params={})
      request "/updateDealItems.php", params
    end

    def send_sale_to_client(params={})
      request "/sendSaleToClient.php", params
    end

    def get_deals(params={})
      raise "amount is required" if params[:amount].nil?
      raise "pageno is required" if params[:pageno].nil?
      request "/getDeals.php", params
    end

    def get_deals_by_contact_or_company(params={})
      raise "contact_or_company is required" if params[:contact_or_company].nil?
      raise "contact_or_company_id is required" if params[:contact_or_company_id].nil?
      request "/getDealsByContactOrCompany.php", params
    end

    def get_deals_by_project(params={})
      raise "project_id is required" if params[:project_id].nil?
      request "/getDealsByProject.php", params
    end

    def get_deal(params={})
      raise "deal_id is required" if params[:deal_id].nil?
      request "/getDeal.php", params
    end

    def get_deal_phase_changes(params={})
      raise "deal_id is required" if params[:deal_id].nil?
      request "/getDealPhaseChanges.php", params
    end

    def get_all_deal_phase_changes(params={})
      raise "date_from is required" if params[:date_from].nil?
      raise "date_to is required" if params[:date_to].nil?
      request "/getAllDealPhaseChanges.php", params
    end

    def get_deal_phases(params={})
      request "/getDealPhases.php", params
    end

    def get_deal_sources(params={})
      request "/getDealSources.php", params
    end

    def add_invoice(params={})
      raise "contact_or_company is required" if params[:contact_or_company].nil?
      raise "contact_or_company_id is required" if params[:contact_or_company_id].nil?
      raise "sys_department_id is required" if params[:sys_department_id].nil?
      request "/addInvoice.php", params
    end

    def update_invoice_payment_status(params={})
      raise "invoice_id is required" if params[:invoice_id].nil?
      raise "status is required" if params[:status].nil?
      raise "status must be 'paid' or 'not_paid'" unless ['paid', 'not_paid'].include?(params[:status])
      request "/setInvoicePaymentStatus.php", params
    end

    def book_draft_invoice(params={})
      raise "invoice_id is required" if params[:invoice_id].nil?
      request "/bookDraftInvoice.php", params
    end

    def update_invoice(params={})
      raise "invoice_id is required" if params[:invoice_id].nil?
      request "/updateInvoice.php", params
    end

    def update_invoice_comments(params={})
      raise "invoice_id is required" if params[:invoice_id].nil?
      raise "comments is required" if params[:comments].nil?
      request "/updateInvoiceComments.php", params
    end

    def delete_invoice(params={})
      raise "invoice_id is required" if params[:invoice_id].nil?
      request "/deleteInvoice.php", params
    end

    def get_invoice(params={})
      raise "invoice_id is required" if params[:invoice_id].nil?
      request "/getInvoice.php", params
    end

    def get_creditnote(params={})
      raise "creditnote_id is required" if params[:creditnote_id].nil?
      request "/getCreditnote.php", params
    end

    def download_invoice_pdf(params={})
      raise "invoice_id is required" if params[:invoice_id].nil?
      request "/downloadInvoicePDF.php", params
    end

    def download_creditnote_pdf(params={})
      raise "creditnote_id is required" if params[:creditnote_id].nil?
      request "/downloadCreditnotePDF.php", params
    end

    def get_invoices(params={})
      raise "date_from is required" if params[:date_from].nil?
      raise "date_to is required" if params[:date_to].nil?
      request "/getInvoices.php", params
    end

    def get_creditnotes(params={})
      raise "date_from is required" if params[:date_from].nil?
      raise "date_to is required" if params[:date_to].nil?
      request "/getCreditnotes.php", params
    end

    def get_invoices_by_project(params={})
      raise "project_id is required" if params[:project_id].nil?
      request "/getInvoicesByProject.php", params
    end

    def get_creditnotes_by_project(params={})
      raise "project_id is required" if params[:project_id].nil?
      request "/getCreditnotesByProject.php", params
    end

    def get_invoices_by_deal(params={})
      raise "deal_id is required" if params[:deal_id].nil?
      request "/getInvoicesByDeal.php", params
    end

    def add_creditnote(params={})
      raise "invoice_id is required" if params[:invoice_id].nil?
      request "/addCreditnote.php", params
    end

    def send_invoice(params={})
      raise "invoice_id is required" if params[:invoice_id].nil?
      raise "email_to is required" if params[:email_to].nil?
      raise "email_subject is required" if params[:email_subject].nil?
      raise "email_text is required" if params[:email_text].nil?
      request "/sendInvoice.php", params
    end

    def get_invoice_reminder_templates(params={})
      raise "invoice_id is required" if params[:invoice_id].nil?
      request "/getInvoiceReminderTemplates.php", params
    end

    def get_invoice_reminder_template_content(params={})
      raise "invoice_id is required" if params[:invoice_id].nil?
      raise "template_id is required" if params[:template_id].nil?
      request "/getInvoiceReminderTemplateContent.php", params
    end

    def get_invoice_cloud_url(params={})
      raise "invoice_id is required" if params[:invoice_id].nil?
      request "/getInvoiceCloudURL.php", params
    end

    def get_bookkeeping_accounts(params={})
      raise "sys_department_id is required" if params[:sys_department_id].nil?
      request "/getBookkeepingAccounts.php", params
    end

    def add_product(params={})
      raise "name is required" if params[:name].nil?
      raise "price is required" if params[:price].nil?
      request "/addProduct.php", params
    end

    def update_product(params={})
      raise "product_id is required" if params[:product_id].nil?
      request "/updateProduct.php", params
    end

    def delete_product(params={})
      raise "product_id is required" if params[:product_id].nil?
      request "/deleteProduct.php", params
    end

    def get_product(params={})
      raise "product_id is required" if params[:product_id].nil?
      request "/getProduct.php", params
    end

    def get_products(params={})
      raise "amount is required" if params[:amount].nil?
      raise "pageno is required" if params[:pageno].nil?
      request "/getProducts.php", params
    end

    def add_ticket(params={})
      raise "subject is required" if params[:subject].nil?
      request "/addTicket.php", params
    end

    def update_ticket(params={})
      raise "ticket_id is required" if params[:ticket_id].nil?
      request "/updateTicket.php", params
    end

    def add_ticket_message(params={})
      raise "ticket_id is required" if params[:ticket_id].nil?
      raise "message is required" if params[:message].nil?
      request "/addTicketMessage.php", params
    end

    def get_tickets(params={})
      raise "type is required" if params[:type].nil?
      request "/getTickets.php", params
    end

    def get_ticket(params={})
      raise "ticket_id is required" if params[:ticket_id].nil?
      request "/getTickets.php", params
    end

    def get_ticket_messages(params={})
      raise "ticket_id is required" if params[:ticket_id].nil?
      raise "include_internal_message is required" if params[:include_internal_message].nil?
      raise "include_third_party_message is required" if params[:include_third_party_message].nil?
      request "/getTicketMessages.php", params
    end

    def get_ticket_message(params={})
      raise "message_id is required" if params[:message_id].nil?
      request "/getTicketMessage.php", params
    end

    def get_ticket_cloud_url(params={})
      raise "ticket_id is required" if params[:ticket_id].nil?
      request "/getTicketCloudURL.php", params
    end

    def add_note(params={})
      raise "object_type is required" if params[:object_type].nil?
      raise "object_id is required" if params[:object_id].nil?
      raise "note_title is required" if params[:note_title].nil?
      request "/addNote.php", params
    end

    def get_notes(params={})
      raise "object_type is required" if params[:object_type].nil?
      raise "object_id is required" if params[:object_id].nil?
      raise "pageno is required" if params[:pageno].nil?
      request "/addNote.php", params
    end

    def add_company(params={})
      raise "name is required" if params[:name].nil?
      request "/addCompany.php", params
    end

    def update_company(params={})
      raise "company_id is required" if params[:company_id].nil?
      raise "track_changes is required" if params[:track_changes].nil?
      request "/updateCompany.php", params
    end

    def delete_company(params={})
      raise "company_id is required" if params[:company_id].nil?
      request "/deleteCompany.php", params
    end

    def get_companies(params={})
      raise "amount is required" if params[:amount].nil?
      raise "pageno is required" if params[:pageno].nil?
      request "/getCompanies.php", params
    end

    def get_company(params={})
      raise "company_id is required" if params[:company_id].nil?
      request "/getCompany.php", params
    end

    def get_business_types(params={})
      raise "country is required" if params[:country].nil?
      request "/getBusinessTypes.php", params
    end

    private

    def request(path, data={})
      headers = {
        "Content-Type" => "application/x-www-form-urlencoded"
      }
      data[:api_group] = @api_group
      data[:api_secret] = @api_secret
      data = URI.encode_www_form data
      response = @http.post "/api" + path, data, headers
      code = Integer(response.code)
      if code >= 200 && code < 300
        begin
          JSON.parse response.body, :symbolize_names => true
        rescue
          response.body
        end
      elsif code == 505
        raise "Error: API rate limit reached"
      else
        err = "unknown error"
        begin
          err = JSON.parse(response.body)["reason"] || "#{response.body}"
        rescue
          err = "#{response.body}"
        end
        raise "HTTP Error #{code}: " + err
      end
    end

    def required_params(attributes, params)
      attributes.each { |attr| raise "#{attr} is required" if params[attr].nil? }
    end
  end
end
